pallets/unicks/src/lib.rs

```diff
@@ -0,0 +1,438 @@
+		/// A unick has been blacklisted.
+		UnickBlacklisted { unick: UnickOf<T> },
+		/// A unick has been unblacklisted.
+		UnickUnblacklisted { unick: UnickOf<T> },
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// The tx submitter does not have enough fund to pay for the deposit.
+		InsufficientFunds,
+		/// The specified unick has already been previously claimed.
+		UnickAlreadyClaimed,
+		/// The specified unick does not exist.
+		UnickNotFound,
+		/// The specified owner already owns a unick.
+		OwnerAlreadyExisting,
+		/// The specified unick has been blacklisted and cannot be interacted
+		/// with.
+		UnickBlacklisted,
+		/// The specified unick is not currently blacklisted.
+		UnickNotBlacklisted,
+		/// The specified unick has already been previously blacklisted.
+		UnickAlreadyBlacklisted,
+		/// The actor cannot performed the specified operation.
+		NotAuthorized,
+		/// A valid unick cannot be decoded from the provided input.
+		InvalidUnickFormat,
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Assign the specified unick to the owner as specified in the origin.
+		///
+		/// The unick must not have already been claimed by someone else and the
+		/// owner must not already own another unick.
+		///
+		/// Emits `UnickClaimed` if the operation is carried out successfully.
+		///
+		/// # <weight>
+		/// Weight: O(1)
+		/// - Reads: Unicks, Owner, Blacklist storage entries + available
+		///   currency check + origin check
+		/// - Writes: Unicks, Owner storage entries + currency deposit reserve
+		/// # </weight>
+		#[pallet::weight(0)]
+		pub fn claim(origin: OriginFor<T>, unick: UnickInput<T>) -> DispatchResult {
+			let origin = T::RegularOrigin::ensure_origin(origin)?;
+			let payer = origin.sender();
+			let owner = origin.subject();
+
+			let decoded_unick = Self::check_claiming_preconditions(unick, &owner, &payer)?;
+
+			// No failure beyond this point
+
+			Self::register_unick(decoded_unick.clone(), owner.clone(), payer);
+			Self::deposit_event(Event::<T>::UnickClaimed {
+				owner,
+				unick: decoded_unick,
+			});
+
+			Ok(())
+		}
+
+		/// Release the provided unick from its owner.
+		///
+		/// The origin must be the owner of the specified unick.
+		///
+		/// Emits `UnickReleased` if the operation is carried out successfully.
+		///
+		/// # <weight>
+		/// Weight: O(1)
+		/// - Reads: Owner storage entry + origin check
+		/// - Writes: Unicks, Owner storage entries + currency deposit release
+		/// # </weight>
+		#[pallet::weight(0)]
+		pub fn release_by_owner(origin: OriginFor<T>, unick: UnickInput<T>) -> DispatchResult {
+			let origin = T::RegularOrigin::ensure_origin(origin)?;
+			let owner = origin.subject();
+
+			let unick = UnickOf::<T>::try_from(unick.into_inner()).map_err(DispatchError::from)?;
+
+			Self::check_releasing_preconditions_for_owner(&unick, &owner)?;
```

Comment:
The case for `chek_claiming_preconditions` is that it takes a unick input as input, and returns an input ready to be stored if the checks fail. On the other hand, all `check_releasing_preconditions` does is to check that the storage state is valid for the given unick. If we build the unick in the function it means that we have to remove it from the storage, right? In that case it would create side effects, which I wanted to avoid inside the check_* functions.

Reply:
> The case for `chek_claiming_preconditions` is that it takes a unick input as input, and returns an input ready to be stored if the checks **fail**. 

Did you mean `succeed`? Let's chat quickly tomorrow. To me it just seemed that we could easily move the 
```
let unick = UnickOf::<T>::try_from(unick.into_inner()).map_err(DispatchError::from)?;
````

inside `check_releasing_preconditions_for_owner` as used in `check_claiming_preconditions`.

Reply:
Makes sense. Part of https://github.com/KILTprotocol/mashnet-node/pull/314/commits/422ec528df34da06dde74153a47b94c4b153d161